### PR TITLE
Remove Deprecations. Replace tangent_transforms functionality

### LINKED ISSUE #113: Remove the use of `nothing` to designate non-differentiability in `rrule_test`
Reusing similar example as in #112:
```julia
function partly_dne(a, i)
    return a[i]
end

function ChainRulesCore.rrule(::typeof(partly_dne), a, i)
    y = partly_dne(a, i)
    function partly_dne_pullback(ȳ)
        grad = zeros(size(a))
        grad[i] = ȳ
        return (NO_FIELDS, grad, DoesNotExist())
    end
    return y, partly_dne_pullback
end
```
It seems un `ChainRules` like that `nothing` should be used instead of `DoesNotExist()`. Should we change this for v1?
```
julia> rrule_test(partly_dne, rand(), (rand(4), rand(4)), (1, nothing))

### LINKED ISSUE #164: remove deprecations for `isapprox` and `f/rrule_test`

### PR DESCRIPTION
**Pull request by mzgubic**
Closes #113, closes #164.

~Do we also want https://github.com/JuliaDiff/ChainRulesTestUtils.jl/issues/179?~

Make sure thunks are supported in rules in
- [x] [SpecialFunctions](https://github.com/JuliaMath/SpecialFunctions.jl/pull/329) 
- [x] [BlockDiagonals](https://github.com/invenia/BlockDiagonals.jl/pull/73)
- [x] [ChainRules](https://github.com/JuliaDiff/ChainRules.jl/pull/449)

### CODE REVIEW

**Comment by oxinabox** [comment]
> Do we also want #179?

No, probably not.

Why are we releasing this right now?
Or are we just getting ready with the PR open?

**Comment by mzgubic** [comment]
Yeah, just making sure we don't forget to deprecate things this time. We are planning to release two breaking versions though, right? - One which requires rules to support thunks (0.8) and one which requires the support for ZeroTangent as well (1.0)?

**Comment by oxinabox** [comment]
> We are planning to release two breaking versions though, right? - One which requires rules to support thunks (0.8) and one which requires the support for ZeroTangent as well (1.0)?

I was thinking just 1.

**Comment by oxinabox** [comment]
looks like we have some issues with it calling `Float64(::Thunk)`
I will debug this later today.

**Comment by oxinabox** [comment]
This is now passing tests.
Because I added the missing unthunks.
Many of them were for `frules`.
I don't think using thunks in `frules` is a good idea because it breaks mutation support.

I think we should stop testing thunks in frules.

**Comment by mzgubic** [comment]
Very naive, but how does it break mutation support? Is it that frules which support mutation can't receive thunks?

**Comment by oxinabox** [comment]
> Very naive q, but how does it break mutation support? Is it that frules which support mutation can't receive thunks?

frules for mutating primal functions can't recieve thunks.
The first rule of mutation support club, is that mutation of the primal needs to result in matching mutation of the tangent.
Thunks are effectively immutable.
In general, for this reason, mutation support is much more fussy about the types involved.

Shall i remove that from this PR?
Probably easiest as I can see more clearly which frules i changed that i should not have

**Comment by mzgubic** [comment]
Thanks, makes sense, I was just wondering whether there was anything else to it.

I assume mutation support is more important, right? So let's remove the default? I can also do that

**Comment by oxinabox** [comment]
>  I assume mutation support is more important, right? So let's remove the default? I can also do that

I will do this

File: test/testers.jl [hunk 1]
```diff
@@ -33,12 +33,12 @@ Base.iterate(f::Foo, state) = iterate(f.a, state)
...
     foo = Foo(a)
     function Foo_pullback(Δfoo)
-        return NoTangent(), Δfoo.a
+        return NoTangent(), unthunk(Δfoo).a
     end
     return foo, Foo_pullback
 end
 function ChainRulesCore.frule((_, Δa), ::Type{Foo}, a)
-    return Foo(a), Foo(Δa)
+    return Foo(a), Foo(unthunk(Δa))
```
Comment (oxinabox) [inline test/testers.jl:41]:
```suggestion
    return Foo(a), Foo(Δa)
```

File: test/testers.jl [hunk 2]
```diff
@@ -49,7 +49,8 @@ function ChainRulesCore.rrule(f::Foo, x)
     end
     return y, Foo_pullback
 end
-function ChainRulesCore.frule((Δf, Δx), f::Foo, x)
+function ChainRulesCore.frule((Δf_, Δx), f::Foo, x)
+    Δf = unthunk(Δf_)
```
Comment (oxinabox) [inline test/testers.jl:53]:
```suggestion
function ChainRulesCore.frule((Δf, Δx), f::Foo, x)
```

File: test/testers.jl [hunk 3]
```diff
@@ -158,7 +159,7 @@ struct MySpecialConfig <: RuleConfig{Union{MySpecialTrait}} end
 
         @testset "check not inferred in frule" begin
             function ChainRulesCore.frule((_, Δx), ::typeof(f_noninferrable_frule), x)
-                return (x, x > 0 ? Float64(Δx) : Float32(Δx))
+                return (x, x > 0 ? Float64(unthunk(Δx)) : Float32(unthunk(Δx)))
```
Comment (oxinabox) [inline test/testers.jl:162]:
```suggestion
                return (x, x > 0 ? Float64(Δx) : Float32(Δx))
```

File: test/testers.jl [hunk 4]
```diff
@@ -233,10 +234,13 @@ struct MySpecialConfig <: RuleConfig{Union{MySpecialTrait}} end
 
         @testset "check non-inferrable primal still passes if pullback inferrable" begin
             function ChainRulesCore.frule((_, Δx), ::typeof(f_inferrable_pullback_only), x)
-                return (x > 0 ? Float64(x) : Float32(x), x > 0 ? Float64(Δx) : Float32(Δx))
+                T  = x > 0 ? Float64 : Float32
+                return T(x), T(unthunk(Δx))
```
Comment (oxinabox) [inline test/testers.jl:238]:
```suggestion
                return T(x), T(Δx)
```

File: test/testers.jl [hunk 5]
```diff
@@ -462,7 +468,8 @@ struct MySpecialConfig <: RuleConfig{Union{MySpecialTrait}} end
             return s
         end
 
-        function ChainRulesCore.frule((_, Δiter), ::typeof(iterfun), iter)
+        function ChainRulesCore.frule((_, Δiter_), ::typeof(iterfun), iter)
+            Δiter = unthunk(Δiter_)
```
Comment (oxinabox) [inline test/testers.jl:472]:
```suggestion
        function ChainRulesCore.frule((_, Δiter), ::typeof(iterfun), iter)
```

**Comment by mzgubic** [comment]
should we test that we catch bad rules with this?

**Comment by mzgubic** [comment]
I approve the commits added (but can't click approve). One thought - should we also keep the empty deprecated.jl file in the tests so that we don't forget to add it?

**Comment by oxinabox** [comment]
>  One thought - should we also keep the empty deprecated.jl file in the tests so that we don't forget to add it?

Maybe, I think it is not a big deal either way, So I am not going to retrigger CI for that

File: docs/src/index.md [hunk 6]
```diff
@@ -72,7 +72,7 @@ The call will test the `rrule` for function `f` at the point `x`, and similarly
 ```jldoctest ex
 julia> test_rrule(two2three, 3.33, -7.77);
 Test Summary:                            | Pass  Total
-test_rrule: two2three on Float64,Float64 |    8      8
+test_rrule: two2three on Float64,Float64 |    9      9
```
[the remark this answers is not in the document]
Comment (oxinabox) [reply docs/src/index.md:75]: too slow little bot, too slow

**Pull request merged**